File: src/main/java/com/grizzly/baseViews/BaseView.java
package com.grizzly.baseViews;

import android.app.Activity;
import android.content.Context;
import android.content.ContextWrapper;
import android.support.annotation.IntDef;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by FcoPardo on 7/6/16.
 */
public class BaseView {

    public static final int PARENT_UNKNOW = 0x0000;
    public static final int PARENT_LISTVIEW = 0x0001;
    public static final int PARENT_RECYCLER = 0x0002;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({PARENT_UNKNOW, PARENT_LISTVIEW, PARENT_RECYCLER})
    public @interface ParentType{}


    public interface OnDataDrivenView<T>{
        void setData(T data);
        T getData();
    }

    static View inflateLayout(int layout, Context context, ViewGroup group){
        return inflateLayout(layout, context, group, true);
    }

    static View inflateLayout(int layout, Context context, ViewGroup group, boolean attachToRoot){
        String infService = Context.LAYOUT_INFLATER_SERVICE;
        LayoutInflater inflater = (LayoutInflater) context.getSystemService(infService);
        return inflater.inflate(layout, group, attachToRoot);
    }

    static Activity getActivity(View view) {
        Context context = view.getContext();
        while (context instanceof ContextWrapper) {
            if (context instanceof Activity) {
                return (Activity)context;
            }
            context = ((ContextWrapper)context).getBaseContext();
        }
        return null;
    }

}

File: src/main/java/com/grizzly/baseViews/BaseView.kt
@file:JvmName("BaseView")
package com.grizzly.baseViews

import android.content.Context
import android.support.annotation.IntDef
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import java.lang.annotation.Retention
import java.lang.annotation.RetentionPolicy

/**
 * Created by FcoPardo on 4/29/18.
 */
class BaseView {

    companion object instance{
        const val PARENT_UNKNOW   : Int = 0x0000
        const val PARENT_LISTVIEW : Int = 0x0001
        const val PARENT_RECYCLER : Int = 0x0002

        @JvmStatic
        internal fun inflateLayout(layout: Int, context: Context, group: ViewGroup): View {
            return inflateLayout(layout, context, group, true)
        }

        @JvmStatic
        internal fun inflateLayout(layout: Int, context: Context, group: ViewGroup, attachToRoot: Boolean): View {
            val infService = Context.LAYOUT_INFLATER_SERVICE
            val inflater = context.getSystemService(infService) as LayoutInflater
            return inflater.inflate(layout, group, attachToRoot)
        }
    }

    @Retention(RetentionPolicy.SOURCE)
    @IntDef(PARENT_UNKNOW, PARENT_LISTVIEW, PARENT_RECYCLER)
    annotation class ParentType

    interface OnDataDrivenView<T> : OnVariableView {
        var data: T
    }

    interface OnVariableView {

        var layout : Int
        var aParentType : Int

        fun ViewGroup.inflateBaseLayout() {
            setContainer()
            if (layout > 0) BaseView.inflateLayout(layout, context, this)
            inflateComponents()
        }

        fun inflateComponents()

        fun setContainer()

        fun getParentType(): Int {
            return aParentType.toInt()
        }

        fun <T : OnVariableView> setParentType(@BaseView.ParentType parentType: Int): T {
            this.aParentType = parentType
            return this as T
        }
    }
}